kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/records/BatchAwareMemoryRecordsBuilder.java

```diff
@@ -225,12 +237,28 @@ public BatchAwareMemoryRecordsBuilder appendWithOffset(long offset, Record recor
     }
 
     /**
-     * Builds the memory records
+     * Builds and returns the memory records.
+     * Calling this multiple times is safe and will return a MemoryRecord with the same content,
+     * <em>assuming the {@code buffer} argument passed to the
+     * {@linkplain #BatchAwareMemoryRecordsBuilder(ByteBufferOutputStream) constructor}
+     * has not been modified or used between calls</em>.
      * @return the memory records
      */
     public @NonNull MemoryRecords build() {
+        boolean needsFlip = builder == null || !builder.isClosed();
         maybeAppendCurrentBatch();
-        return MemoryRecords.readableRecords(buffer.buffer().flip());
+        ByteBuffer buffer = this.buffer.buffer();
+        if (needsFlip) {
+            buffer.flip();
+        }
+        return MemoryRecords.readableRecords(buffer);
```

Comment:
don't hide the field.

```suggestion
        ByteBuffer buf = this.buffer.buffer();
        if (needsFlip) {
            buf.flip();
        }
        return MemoryRecords.readableRecords(buf);
```